### PR TITLE
Prevent DeepPartial from recursively deep partialing primitive values.

### PR DESCRIPTION
**Pull request by threehams**
DeepPartial currently handles arrays and readonly arrays, but ends up applying `DeepPartial` recursively to primitives until TS eventually detects the loop and forces the properties to `unknown`, which leads to some _very_ confusing types coming out of `getValues` when using `useForm` without a type parameter.

```typescript
const { getValues } = useForm({ defaultValues: { name: "" } })
getValues().name;
(property) test: {
    toString: unknown;
    charAt: unknown;
    charCodeAt: unknown;
    concat: unknown;
    indexOf: unknown;
    lastIndexOf: unknown;
    localeCompare: unknown;
    match: unknown;
    replace: unknown;
    search: unknown;
    slice: unknown;
    split: unknown;
    substring: unknown;
    toLowerCase: unknown;
    toLocaleLowerCase: unknown;
    toUpperCase: unknown;
    toLocaleUpperCase: unknown;
    ... 30 more ...;
    matchAll: unknown;
}
```

```
Type '{ toString: unknown; charAt: unknown; charCodeAt: unknown; concat: unknown; indexOf: unknown; lastIndexOf: unknown; localeCompare: unknown; match: unknown; replace: unknown; search: unknown; slice: unknown; split: unknown; substring: unknown; toLowerCase: unknown; toLocaleLowerCase: unknown; toUpperCase: unknown; to...' is not assignable to type 'string'.ts(2322)
```

This PR stops that by only applying `DeepPartial` to objects (anything that extends `{ [key: string]: unknown }`.

### CODE REVIEW

**Comment by bluebill1049** [comment]
wow! thank you very much @threehams @kotarella1110 you want to take a quick look on this one.

**Comment by kotarella1110** [comment]
@bluebill1049 It seems that unit tests are not running on the CI.
@threehams I cloned your repository and tried unit tests, but it failed...

**Comment by bluebill1049** [comment]
hmmm strange, i will clone it and check it out.

**Comment by bluebill1049** [comment]
i think this will fixes it @kotarella1110 https://github.com/react-hook-form/react-hook-form/commit/b8793f504db7bbf9d11bfb27c4ac7439b8df31ea

**Comment by kotarella1110** [comment]
@bluebill1049 Thanks!

**Comment by bluebill1049** [comment]
good catch and testing @kotarella1110 👍

**Comment by threehams** [comment]
hmmm. Working on tests, and looking into why types/runtime aren't matching up.

File: src/useForm.test.tsx [hunk 1]
```diff
@@ -1431,5 +1431,23 @@ describe('useForm', () => {
...
+              values: '5',
+            },
+          },
+        }),
+      );
+      const test: string = result.current.getValues().test;
+      expect(test).toEqual('data');
+      const deep: { values: string } = result.current.getValues().deep;
+      expect(deep).toEqual({ values: '5' });
+    });
```
Comment (kotarella1110) [inline src/useForm.test.tsx:1451]:
```suggestion

    it('should infer from defaultValues without a type parameter', () => {
      const { result } = renderHook(() =>
        useForm<{ test: string; deep: { values: string } }>({
          mode: VALIDATION_MODE.onSubmit,
          defaultValues: {
            test: 'data',
            deep: {
              values: '5',
            },
          },
        }),
      );

      (getFieldsValues as any).mockImplementation(async () => {
        return {};
      });

      act(() => {
        result.current.register({ type: 'text', name: 'test' });
        result.current.register({ type: 'text', name: 'deep.nested' });
      });

      act(() => {
        const test: string = result.current.getValues({ nest: true }).test;
        expect(test).toEqual('data');
        const deep: { values: string } = result.current.getValues({
          nest: true,
        }).deep;
        expect(deep).toEqual({ values: '5' });
      });
    });
```

Comment (kotarella1110) [inline src/useForm.test.tsx:1451]:
```suggestion

    it('should infer from defaultValues without a type parameter', () => {
      const { result } = renderHook(() =>
        useForm<{ test: string; deep: { values: string } }>({
          mode: VALIDATION_MODE.onSubmit,
          defaultValues: {
            test: 'data',
            deep: {
              values: '5',
            },
          },
        }),
      );

      (getFieldsValues as any).mockImplementation(async () => {
        return {};
      });

      act(() => {
        const test: string = result.current.getValues({ nest: true }).test;
        expect(test).toEqual('data');
        const deep: { values: string } = result.current.getValues({
          nest: true,
        }).deep;
        expect(deep).toEqual({ values: '5' });
      });
    });
```

**Comment by threehams** [comment]
Thanks, I was struggling with figuring out how to unit test this. I'm used to making a mock component within the test and asserting the DOM output.

**Review by bluebill1049 — DISMISSED** [review]

File: src/useForm.test.tsx [hunk 2]
```diff
@@ -1431,5 +1431,32 @@ describe('useForm', () => {
         } as React.SyntheticEvent);
       });
     });
+
+    it('should infer from defaultValues without a type parameter', async () => {
```
Comment (bluebill1049) [inline src/useForm.test.tsx:1435]:
```suggestion
    it('should infer from defaultValues without a type parameter', () => {
```

Comment (bluebill1049) [reply src/useForm.test.tsx:1435]: don't think we need `async` here

Comment (threehams) [reply src/useForm.test.tsx]: nope. There's [a useful rule for it](https://eslint.org/docs/rules/require-await) if you're interested.

**Review by bluebill1049 — APPROVED** [review]

**Review by kotarella1110 — APPROVED** [review]
LGTM 🎊

**Comment by bluebill1049** [comment]
thank you for your contribution and @kotarella1110's awesome pick up!

**Pull request merged**